### ISSUE TITLE
use IfNotPresent in imagePullPolicy of manifests

### ISSUE THREAD

**Issue by daengdaengLee** [argoproj/argo-cd#20252]
# Summary

in manifests/install.yaml,
AS-IS : `imagePullPolicy: Always`
TO-BE : `imagePullPolicy: IfNotPresent`

# Motivation

In some cases, it could be hard(or troublesome) to connect k8s and image registry. (for many reasons)
I think if someone want to load ArgoCD image manually to k8s cluster and use that image, `IfNotPresent` option will be helpful.

# Proposal

Change `Always` -> `IfNotPresent`

# Note

If this proposal is accepted, I will make the code changes myself and submit a PR.
Thanks!

**Comment by pasha-codefresh** [comment]
I think it is good idea, dont see any reason why we need always. But we need resolve issue with latest tag first, which we use in master branch. This change will break some behavior

**Comment by daengdaengLee** [comment]
Hello, @pasha-codefresh.

1. Could you explain what behavior it will break and how it associated with latest tag, please? (Sorry, I'm not good at k8s setup)
2. Is there an issue or discussion about the latest tag in master? 

Thanks!

**Comment by pasha-codefresh** [comment]
See https://stackoverflow.com/questions/74006353/difference-between-always-and-ifnotpresent-imagepullpolicy

When you use an image WITHOUT a tag, Kubernetes will assume that you want the latest version of the image, which is identified by the latest tag by default. If you have multiple versions of the same image in your repository with different tags, such as img1:1.0.0, img1:1.1.0, and img1:latest, Kubernetes will use the image with the tag specified in the pod specification.

If you use IfNotPresent and the image with the specified tag is already present on the worker node, Kubernetes will use that image to start the container, even if there is a newer version of the image available in the repository with the same tag.


Thats mean that if user use 'latest' tag in his manifests, it will stop pull new latest tags

2. No, here no open issues for it as i remember

**Comment by daengdaengLee** [comment]
Hello, @pasha-codefresh.

I understand your explanation.

How about generating the manifest differently based on the conditions in `hack/update-manifests.sh` ?
* if the image tag == latest -> Use `Always`
* else -> Use `IfNotPresent`

I'm not sure it's a good idea. If I missed anything, please let me know.

Thanks!

**Comment by pasha-codefresh** [comment]
Yes, it can be an option. But maybe if you need such scenario, it will be easier if you modify manifests with kustomize by your own ?

**Comment by crenshaw-dev** [comment]
The [init release](https://github.com/argoproj/argo-cd/blob/master/.github/workflows/init-release.yaml) workflow bumps image tags. We could modify it to also set the imagePullPolicy.

Unfortunately this doesn't get us an image digest, because that's not generated until the next step, when we actually build the image.

**Comment by daengdaengLee** [comment]
As @pasha-codefresh suggested, it might be a good idea to use kustomize to create my own manifests directly.
Or we can add one more option in `./hack/update-manifests.sh` like
```sh
# ...
IMAGE_PULL_POLICY="${IMAGE_PULL_POLICY:-}"

if [ "$IMAGE_PULL_POLICY" = "" ]; then
  IMAGE_PULL_POLICY=Always
fi
# ...

# use IMAGE_PULL_POLICY in kustomize...
```
It can be useful for many people ( including me :) )
We could set the default to preserve the existing behavior to avoid introducing any breaking changes.

It could also be utilized in the ArgoCD pipeline if required later on.

Thanks for comments, @pasha-codefresh, @crenshaw-dev!

**Comment by crenshaw-dev** [comment]
imo it's a good change to make in the release branches for everyone. I've often wondered why we're using pull policy always.

**Comment by daengdaengLee** [comment]
Great! Then I’ll work on it and open a PR!

**Comment by rumstead** [comment]
Just to throw out what I put in slack. 

It is considered good security practice. It forces the container runtime to validate that the cached image on the node is the same one expected from the registry. Always doesn't mean pull "always" it just means verify the digest if the image is on the node.
https://kubernetes.io/docs/concepts/containers/_print/#image-pull-policy

**Comment by warjiang** [comment]
> Just to throw out what I put in slack.
> 
> It is considered good security practice. It forces the container runtime to validate that the cached image on the node is the same one expected from the registry. Always doesn't mean pull "always" it just means verify the digest if the image is on the node. https://kubernetes.io/docs/concepts/containers/_print/#image-pull-policy

I'm in favor of use 'IfNotPresent' as the default imagePullPolicy
Even if it only check the digest, it will not be good for some offline scenarios. I think if we ready want to ensure the latest image, we can use the digest of image like: 'https://quay.io/argoproj/argocd:v2.12.6@{sha256 digest}'

**Comment by andrii-korotkov-verkada** [comment]
Maybe I'm missing something here, but why do we use `latest` tag instead of the fixed tag, which we update when doing a release?

**Comment by daengdaengLee** [comment]
Hello, @andrii-korotkov-verkada .


I’m not entirely sure about the reasons in the past, but here is my organized thought based on the discussion regarding the issue mentioned above:

During the development phase, we continuously build new images with the latest tag after development or modifications. At this stage, using the combination of the latest tag and the Always imagePullPolicy allows Kubernetes to always check the hash of the latest tag image in the image repository. If the hash differs from the local image, k8s re-downloads the image. This approach enables us to easily apply and test new images during development without manually deleting previously built images loaded in k8s.

Because we’ve been using this behavior (maybe), I submitted a PR to maintain the latest + Always combination in the master branch while making it easy to switch the imagePullPolicy to IfNotPresent in the release branch by adding/modifying scripts.

@pasha-codefresh , @crenshaw-dev  If there are any inaccuracies in my explanation, please let me know in the comments.


Thanks for reading!
